### PR TITLE
Enable the RequestScope on gRPC service invocations

### LINKED ISSUE #13356: Activate the RequestScope when invoking gRPC services
**Describe the bug**

gRPC services should have the request scope enabled as, without it, they cannot use Hibernate Reactive. 

Note that the `EnableRequestScope` annotation does not work in this case as it closes the request scope **before** the Uni / Multi gets the result. 

**Expected behavior**

Service implementation should have the request scope enabled.

**To Reproduce**

Implement a Hello gRPC service and try accessing Hibernate Reactive session from within the method in a delayed (after the return of the method) manner.

I believe it would need a gRPC interceptor (or maybe a plain CDI interceptor).

### PR DESCRIPTION
**Pull request by cescoffier**
Wrap gRPC service invocation with an interceptor enabling / disabling the request scope. 
The scope is deactivated when the communication is terminated (on cancellation, failure or completion)

Fix #13356 ﻿

### CODE REVIEW

**Comment by gsmet** [comment]
Rebased to get CI fix.

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/RequestScopeHandlerInterceptor.java
Comment (michalszynkiewicz) [inline extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/RequestScopeHandlerInterceptor.java:25]: why is this variable called `managed`?

**Review by michalszynkiewicz — APPROVED** [review]

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/RequestScopeHandlerInterceptor.java
Comment (cescoffier) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/RequestScopeHandlerInterceptor.java:25]: because I suck at naming. It just means that the request scope is "managed" by the interceptor (and not be enabled beforehand)

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/RequestScopeHandlerInterceptor.java
Comment (mkouba) [inline extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/RequestScopeHandlerInterceptor.java:35]: @cescoffier Do you expect this to be called from a different thread than the one that called `reqContext.activate()`?

**Review by gsmet — DISMISSED** [review]
Let's wait for @cescoffier to answer @mkouba 's question before merging.

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/RequestScopeHandlerInterceptor.java
Comment (cescoffier) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/RequestScopeHandlerInterceptor.java:35]: It could be the case if the method is annotated with @Blocking or if the implementation calls "onComplete" or "onError" from another thread.

Comment (mkouba) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/RequestScopeHandlerInterceptor.java:35]: Well, I don't have enough knowledge about gRPC and its interceptors but `deactivate()` should be always called on the same thread (to cleanup the thread locals) and we should also call `destroy()` to destroy the beans correctly when the request processing is over. In reactive routes we capture the state and add an end handler: https://github.com/quarkusio/quarkus/blob/master/extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/RouteHandler.java#L46-L52

Comment (cescoffier) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/RequestScopeHandlerInterceptor.java:35]: Hum... that is a bit tricky, I need another approach as we cannot guarantee this.
Even if I capture the context, that context may be used by another invocation, which may even complete before the first call (and so destroy the beans created for the first call). 

There is no "routing context" in gRPC which makes things a bit more tricky. 

I'm wondering how you handle the context reuse in reactive routes?

**Comment by cescoffier** [comment]
Converting PR to draft as the approach does not work as we may not be on the same thread.

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/RequestScopeHandlerInterceptor.java
Comment (mkouba) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/RequestScopeHandlerInterceptor.java:35]: > I'm wondering how you handle the context reuse in reactive routes?

What exactly do you mean? In any case, the question should be "how do we handle the context reuse..." ;-)

Comment (cescoffier) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/RequestScopeHandlerInterceptor.java:35]: If I understand correctly, the request scope is thread dependant, which means that it stored in a thread-local attached to the thread "activating" it. As a consequence, deactivation should be done from the same thread, so clean up everything.

Everything work when the thread is used to handle a single concurrent call. 
But it may not be the case in the reactive world.

For example in the reactive routes, your method is called on the IO thread (event loop). That thread may be used to handle multiple concurrent requests. If one of these requests enables the request scope, what does it means for the other requests handled (concurrently) on the same event loop?  If two concurrent requests both enable the request scope, and one terminates before the other one, how can you be sure that it does not close the request scope of the other (still progressing) request? 

I just found a way to have something similar to the routing context in gRPC, but I need to be sure that we would not step on other's toes.

Comment (mkouba) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/RequestScopeHandlerInterceptor.java:35]: > If I understand correctly, the request scope is thread dependant, which means that it stored in a thread-local attached to the thread "activating" it.

Yes.

> But it may not be the case in the reactive world.

Yes, that's why MP Context Propagation was created (although it does not cover all possible use cases).

> For example in the reactive routes, your method is called on the IO thread (event loop). That thread may be used to handle multiple concurrent requests.

The request context is deactivated (thread local removed) right after a route method completes so it should not be a problem (a IO thread may not execute multiple route methods at the same time). If you do some async inside the route method MP CP makes a snapshot of the context before the async action is performed and this snapshot is propagated further.

Comment (cescoffier) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/RequestScopeHandlerInterceptor.java:35]: Ok, so it requires MP Context propagation in these cases, or you got things messed up. 

So, I should be able to use the same machinery and restore the right (execution) context. I just need to store it and propagate it.

Comment (mkouba) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/RequestScopeHandlerInterceptor.java:35]: > Ok, so it requires MP Context propagation in these cases, or you got things messed up.

Yes, if you need the context to be propagated. In fact, you can leverage the ArC APIs directly but that's not something I would recommend ;-).

> So, I should be able to use the same machinery...

I think so.

Comment (cescoffier) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/RequestScopeHandlerInterceptor.java]: I'm happy to say that I've found a simple solution -> Just capturing the context.

**Comment by gsmet** [comment]
@cescoffier should this one be taken out of draft and reviewed? If so, could you ping the reviewers again once done?

**Comment by cescoffier** [comment]
@gsmet thanks for reminding me about that one! First, let me rebase it.

**Comment by gsmet** [comment]
I would prefer if @mkouba could have a look at this one. @mkouba could you have a look tomorrow morning?

**Review by mkouba — APPROVED** [review]
LGTM.

**Pull request merged**
